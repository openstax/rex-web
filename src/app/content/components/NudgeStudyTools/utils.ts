import { HTMLElement } from '@openstax/types/lib.dom';
import * as Cookies from 'js-cookie';
import React, { useState } from 'react';
import { useSelector } from 'react-redux';
import { useDebouncedWindowSize, useOnScrollTopOffset } from '../../../reactUtils';
import { assertDocument, remsToPx } from '../../../utils';
import { page as pageSelector } from '../../selectors';
import { hasStudyGuides } from '../../studyGuides/selectors';
import { Page } from '../../types';
import {
  arrowDesktopHeight,
  arrowRightMargin,
  arrowTopMargin,
  closeButtonDistanceFromContent,
  closeButtonSize,
  contentMarginLeft,
  contentMarginTop,
  contentWidth,
  cookieNudge,
  daysUntilCookieExpires,
  mobileNudgeStudyToolsTargetId,
  mobileSpotlightPadding,
  nudgeStudyToolsMinPageLimit,
  nudgeStudyToolsShowLimit,
  nudgeStudyToolsTargetId,
  timeIntervalBetweenShowingNudgeInMs,
} from './constants';

interface Positions {
  arrowLeft: number;
  arrowTopOffset: number;
  closeButtonLeft: number;
  closeButtonTopOffset: number;
  contentWrapperLeft: number;
  contentWrapperTopOffset: number;
  spotlightHeight: number;
  spotlightLeftOffset: number;
  spotlightTopOffset: number;
  spotlightWidth: number;
}

export const getPositions = (target: HTMLElement, isMobile: boolean): Positions => {
  const padding = remsToPx(mobileSpotlightPadding);
  const { top, left, height, width } = getBoundingRectOfNudgeTarget(target);
  const spotlightLeftOffset = left - (isMobile ? padding : 0);
  const spotlightWidth = width + (isMobile ? padding * 2 : 0);
  const arrowTopOffset = top + height + remsToPx(arrowTopMargin);
  // left edge of arrow image should be on the middle of the spotlight (adjusted for a margin)
  const arrowLeft = spotlightLeftOffset + (spotlightWidth / 2) - remsToPx(arrowRightMargin);

  const contentWrapperTopOffset = arrowTopOffset
    + remsToPx(arrowDesktopHeight)
    + remsToPx(contentMarginTop);
  const contentWrapperLeft = spotlightLeftOffset + spotlightWidth + remsToPx(contentMarginLeft);
  const closeButtonLeft = contentWrapperLeft + remsToPx(contentWidth) + remsToPx(closeButtonDistanceFromContent);
  const closeButtonTopOffset = contentWrapperTopOffset
    - remsToPx(closeButtonSize)
    - remsToPx(closeButtonDistanceFromContent);

  return {
    arrowLeft,
    arrowTopOffset,
    closeButtonLeft,
    closeButtonTopOffset,
    contentWrapperLeft,
    contentWrapperTopOffset,
    spotlightHeight: height,
    spotlightLeftOffset,
    spotlightTopOffset: top,
    spotlightWidth,
  };
};

// Target may have display set to `contents` so we are calculatings rect for children of passed target.
const getBoundingRectOfNudgeTarget = (target: HTMLElement) => {
  const { top, bottom, left, right } = Array.from(target.children).reduce((rects, child) => {
    const rect = child.getBoundingClientRect();
    rects.top = Math.min(rects.top, rect.top);
    rects.bottom = Math.max(rects.bottom, rect.bottom);
    rects.left = Math.min(rects.left, rect.left);
    rects.right = Math.max(rects.right, rect.right);
    return rects;
    // Starting values depends on if we checks for Math.max or Math.min
  }, { top: 9999, bottom: 0, left: 9999, right: 0 });

  return {
    height: bottom - top,
    left,
    right,
    top,
    width: right - left,
  };
};

export const usePositions = (isMobile: boolean) => {
  const [windowWidth] = useDebouncedWindowSize();
  const scrollTopOffset = useOnScrollTopOffset();
  const [positions, setPositions] = React.useState<Positions | null>(null);
  const studyGuides = useSelector(hasStudyGuides);

  React.useEffect(() => {
    const document = assertDocument();
    const target = document.querySelector(
      `#${isMobile ? mobileNudgeStudyToolsTargetId : nudgeStudyToolsTargetId}`
    ) as HTMLElement | null;
    if (target) {
      // Make sure that we calculate positions with body overflow set to hidden
      // because it causes scrollbar to hide which results in different positions.
      const prevOverflow = document.body.style.overflow;
      document.body.style.overflow = 'hidden';
      setPositions(getPositions(target, isMobile));

      // Resets to the value from before calculations. We want this style change to be handled
      // directly in the component.
      document.body.style.overflow = prevOverflow;
    }
    return () => setPositions(null);
  }, [scrollTopOffset, studyGuides, windowWidth, isMobile]);

  return positions;
};

export const getCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.counter) || 0);
};

export const getDateCookie = () => {
  const lastShownDate = Cookies.get(cookieNudge.date);
  return lastShownDate ? new Date(lastShownDate) : undefined;
};

export const passedTimeInterval = () => {
  const lastShownDate = getDateCookie();
  return !lastShownDate
    ? true
    : (Date.now() - lastShownDate.getTime()) > timeIntervalBetweenShowingNudgeInMs
  ;
};

export const getPageCounterCookie = () => {
  return Number(Cookies.get(cookieNudge.pageCounter) || 0);
};

export const incrementPageCounterCookie = () => {
  const counter = getPageCounterCookie();
  const newValue = counter + 1;
  Cookies.set(cookieNudge.pageCounter, newValue.toString(), {expires: daysUntilCookieExpires});
  return newValue;
};

export const shouldDisplayNudgeStudyTools = (): boolean => {
  const counter = getCounterCookie();
  const numberOfPagesOpenedByUser = getPageCounterCookie();

  return counter < nudgeStudyToolsShowLimit
    && passedTimeInterval()
    && (numberOfPagesOpenedByUser >= nudgeStudyToolsMinPageLimit);
};

// Set required cookies and reset opened page counter
export const setNudgeStudyToolsCookies = () => {
  const now = new Date();
  const counter = getCounterCookie();
  Cookies.set(cookieNudge.counter, (counter + 1).toString(), {expires: daysUntilCookieExpires});
  Cookies.set(cookieNudge.date, now.toString(), {expires: daysUntilCookieExpires});
  Cookies.remove(cookieNudge.pageCounter);
};

export const useIncrementPageCounter = () => {
  const page = useSelector(pageSelector);
  const [counter, updateCounter] = useState<number>(getPageCounterCookie());
  const pageRef = React.useRef<Page>();

  React.useEffect(() => {
    if (page && page !== pageRef.current && counter <= nudgeStudyToolsMinPageLimit) {
      pageRef.current = page;
      updateCounter(incrementPageCounterCookie());
    }
  }, [page, updateCounter, counter]);

  return counter;
};
