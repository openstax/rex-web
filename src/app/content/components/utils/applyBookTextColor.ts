import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { BookWithOSWebData } from '../../types';

export const applyBookTextColor = (props: {colorSchema: BookWithOSWebData['theme'] | undefined }) =>
  props.colorSchema && css`
    color: ${theme.color.primary[props.colorSchema].foreground};
  `;
