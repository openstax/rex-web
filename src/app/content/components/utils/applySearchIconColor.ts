import { css } from 'styled-components/macro';
import theme from '../../../theme';
import { toolbarIconColor } from '../constants';

const primaryColors = theme.color.primary;

export const applySearchIconColor = (colorSchema: keyof typeof primaryColors | null) =>
  colorSchema ? css`
    color: ${primaryColors[colorSchema].foreground};

    :hover,
    :focus {
        color: ${primaryColors[colorSchema].foreground};
        transition: color 200ms;

        ::before {
            content:"";
            display: block;
            height: 100%;
            position: absolute;
            top: 0;
            right: 0;
            min-width: 45px;
            background-color: rgba(0, 0, 0, 0.06)
            ;
        }
    }
  ` : css`
  color: ${toolbarIconColor.base};
`;
