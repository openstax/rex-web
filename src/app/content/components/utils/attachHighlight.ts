import Highlighter, { Highlight, SerializedHighlight } from '@openstax/highlighter';
import Sentry, { Severity } from '../../../../helpers/Sentry';

const attachHighlight = <T extends Highlight | SerializedHighlight>(
  highlight: T,
  highlighter: Highlighter,
  errorMsg?: (highlight: T) => string
): Highlight | undefined => {
  let attachedHighlight: Highlight | undefined;

  try {
    highlighter.highlight(highlight);
    attachedHighlight = highlighter.getHighlight(highlight.id);
  } catch (e) {
    Sentry.captureException(e);
  }

  if (!attachedHighlight || !attachedHighlight.isAttached()) {
    Sentry.captureException(
      new Error(errorMsg ?  errorMsg(highlight) : `Highlight with id: ${highlight.id} has not been attached.`),
      Severity.Warning
    );
    attachedHighlight = undefined;
  }

  return attachedHighlight;
};

export default attachHighlight;
