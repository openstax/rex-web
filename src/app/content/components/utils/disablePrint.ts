import { css } from 'styled-components';

export const disablePrint = css`@media print { display: none; }`;
