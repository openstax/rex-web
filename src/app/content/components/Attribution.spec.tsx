import cloneDeep from 'lodash/fp/cloneDeep';
import createTestStore from '../../../test/createTestStore';
import { book, page, shortPage } from '../../../test/mocks/archiveLoader';
import { mockCmsBook } from '../../../test/mocks/osWebLoader';
import { reactAndFriends, resetModules } from '../../../test/utils';
import { AppState, Store } from '../../types';
import * as actions from '../actions';
import { initialState } from '../reducer';
import { formatBookData } from '../utils';

// tslint:disable:variable-name
describe('Attribution', () => {
  let React: ReturnType<typeof reactAndFriends>['React'];
  let ReactDOM: ReturnType<typeof reactAndFriends>['ReactDOM'];
  let renderer: ReturnType<typeof reactAndFriends>['renderer'];
  let renderToDom: ReturnType<typeof reactAndFriends>['renderToDom'];
  let TestContainer: ReturnType<typeof reactAndFriends>['TestContainer'];
  let scrollTo: any;

  beforeEach(() => {
    jest.resetAllMocks();
    resetModules();
    const mockScrollTo = scrollTo = jest.fn();
    jest.mock('scroll-to-element', () => mockScrollTo);
    ({React, ReactDOM, renderer, renderToDom, TestContainer} = reactAndFriends());
  });

  describe('in browser', () => {
    let state: AppState;
    let store: Store;
    let Attribution: any;

    beforeEach(() => {
      Attribution = require('./Attribution').default;
      state = (cloneDeep({
        content: {
          ...initialState,
          book: formatBookData(book, mockCmsBook),
          page,
        },
        navigation: { pathname: '/cool path name' },
      }) as any) as AppState;

      store = createTestStore(state);
    });

    const render = () => <TestContainer store={store}>
      <Attribution />
    </TestContainer>;

    it('removes listener when it unmounts', () => {
      const { root, node } = renderToDom(render());
      const details = node;

      details.removeEventListener = jest.fn();

      ReactDOM.unmountComponentAtNode(root);

      if (details) {
        expect(details.removeEventListener).toHaveBeenCalled();
      } else {
        expect(details).toBeTruthy();
      }
    });

    it('scrolls to details when toggled', async() => {
      const { node } = renderToDom(render());
      const details = node;

      if (!document) {
        return expect(document).toBeTruthy();
      }

      details.setAttribute('open', '');

      // wait for dom events to do their thing
      await new Promise((resolve) => setTimeout(resolve, 1));

      expect(scrollTo).toHaveBeenCalledWith(details, expect.anything());
    });

    it('closes attribution on navigation', async() => {
      const { node } = renderToDom(render());
      const details = node;

      if (!document) {
        return expect(document).toBeTruthy();
      }

      details.setAttribute('open', '');

      store.dispatch(actions.receivePage({...shortPage, references: []}));

      expect(details.getAttribute('open')).toBe(null);
    });

    it('doesn\'t close attribution when updating with the same page', async() => {
      const { node } = renderToDom(render());
      const details = node;

      if (!document) {
        return expect(document).toBeTruthy();
      }

      details.setAttribute('open', '');

      store.dispatch(actions.receiveBook(formatBookData(book, mockCmsBook)));

      expect(details.getAttribute('open')).toBe('');

      // clear scrollTo that happens on toggle before finishing the test
      // or there are errors
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    it('mounts and unmounts without a dom', () => {
      const element = renderer.create(render());

      expect(element.unmount).not.toThrow();
    });

    it('throws if cms doesn\'t return a publication date ', () => {
      const newState = (cloneDeep({
        content: {
          ...initialState,
          book: formatBookData(book, { ...mockCmsBook, publish_date: null} ),
          page,
        },
        navigation: { pathname: 'cool path name' },
      }) as any) as AppState;

      store = createTestStore(newState);

      // We are testing if assertNotNull function is throwing error
      // and when it does then react makes a console.error with additional info.
      const spyConsoleError = jest.spyOn(console, 'error')
        .mockImplementationOnce(jest.fn)
      ;

      let message: string | undefined;

      try {
        renderer.create(render());
      } catch (e) {
        message = e.message;
      }

      expect(spyConsoleError).toHaveBeenCalledTimes(1);
      expect(message).toEqual('BUG: Could not find publication date');
    });

    it('displays first two contributing authors when no senior authors were found', () => {
      const newState = (cloneDeep({
        content: {
          ...initialState,
          book: formatBookData(book, { ...mockCmsBook,
            authors: [
              {value: {name: 'Jhon Doe', senior_author: false}},
              {value: {name: 'Jonny Doe', senior_author: false}},
            ],
          }),
        },
      }) as any) as AppState;

      store = createTestStore(newState);
      const { node: details } = renderToDom(render());

      expect(details.children[1].innerHTML).toMatch(`Authors: Jhon Doe, Jonny Doe`);
    });

    it('renders default attribution', async() => {
      store.dispatch(
        actions.receiveBook({...formatBookData(book, mockCmsBook)})
      );
      const component = renderer.create(render());
      expect(component).toMatchSnapshot();
    });

    it('renders special licensing and attribution for HS Physics', async() => {
      store.dispatch(
        actions.receiveBook({...formatBookData(book, mockCmsBook), id: 'cce64fde-f448-43b8-ae88-27705cceb0da'})
      );
      const component = renderer.create(render());
      expect(component).toMatchSnapshot();
    });

    it('renders special licensing and attribution for Statistics', async() => {
      store.dispatch(
        actions.receiveBook({...formatBookData(book, mockCmsBook), id: '394a1101-fd8f-4875-84fa-55f15b06ba66'})
      );
      const component = renderer.create(render());
      expect(component).toMatchSnapshot();
    });

    it('renders special licensing and attribution for Intellectual Property', async() => {
      store.dispatch(
        actions.receiveBook({...formatBookData(book, mockCmsBook), id: '1b4ee0ce-ee89-44fa-a5e7-a0db9f0c94b1'})
      );
      const component = renderer.create(render());
      expect(component).toMatchSnapshot();
    });
  });
});
