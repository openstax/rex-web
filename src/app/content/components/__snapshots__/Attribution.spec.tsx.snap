// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Attribution in browser renders default attribution 1`] = `
<details
  className="Details-xi326o-3 Attribution__AttributionDetails-sc-11isnv6-4 ecFdAU"
  data-analytics-region="attribution"
  data-testid="attribution-details"
>
  <summary
    aria-label="Citation/Attribution"
    className="Attribution__AttributionSummary-sc-11isnv6-2 ccMXsL Details__Summary-xi326o-2 goqLfs"
  >
    <svg
      aria-hidden="true"
      className="Attribution__SummaryClosedIcon-sc-11isnv6-0 gSPJtD Details__ExpandIcon-xi326o-0 cVuUUi StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 192 512"
    >
      <path
        d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        fill="currentColor"
      />
    </svg>
    <svg
      aria-hidden="true"
      className="Attribution__SummaryOpenIcon-sc-11isnv6-1 kRXAve Details__CollapseIcon-xi326o-1 jsdrhW StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 320 512"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
    <span>
      Citation/Attribution
    </span>
  </summary>
  <div
    className="Attribution__Content-sc-11isnv6-3 SNfN"
    dangerouslySetInnerHTML={
      Object {
        "__html": "
<p>
  Want to cite, share, or modify this book? This book is
  <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"http://example.com\\">
    Sweet License 1.0
  </a> and you must attribute OpenStax.
</p>

<strong>Attribution information</strong>
<ul>
  <li>
    If you are redistributing all or part of this book in a print format,
    then you must include on every physical page the following attribution:
    <blockquote>
      Access for free at https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units
    </blockquote>
  </li>
  <li>
    If you are redistributing all or part of this book in a digital format,
    then you must include on every digital page view the following attribution:
    <blockquote>
      Access for free at <a href=\\"https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
    </blockquote>
  </li>
</ul>

<strong>Citation information</strong>
<ul>
  <li>
    Use the information below to generate a citation. We recommend using a
    citation tool such as
    <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"https://www.lib.ncsu.edu/citationbuilder/#/default/default\\">this one</a>.
    <ul>
      <li>
        Authors: Bam Bammerson
      </li>
      <li>
        Publisher/website: OpenStax
      </li>
      <li>
        Book title: Test Book 1
      </li>
      <li>
        Publication date: Jun 21, 2012
      </li>
      <li>
        Location: Houston, Texas
      </li>
      <li>
        Book URL: <a href=\\"https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/book-slug-1/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
      </li>
      <li>
        Section URL: <a href=\\"https://openstax.org/cool path name\\">https://openstax.org/cool path name</a>
      </li>
    </ul>
  </li>
</ul>

<p>
  © Jun 21, 2012 OpenStax. Textbook content produced by OpenStax is licensed under a Sweet License 1.0 license. <strong>The OpenStax name, OpenStax logo, OpenStax book covers, OpenStax CNX name, and OpenStax CNX logo
  are not subject to the Creative Commons license and may not be reproduced without the prior and express written
  consent of Rice University.</strong>
</p>
",
      }
    }
  />
</details>
`;

exports[`Attribution in browser renders special licensing and attribution for HS Physics 1`] = `
<details
  className="Details-xi326o-3 Attribution__AttributionDetails-sc-11isnv6-4 ecFdAU"
  data-analytics-region="attribution"
  data-testid="attribution-details"
>
  <summary
    aria-label="Citation/Attribution"
    className="Attribution__AttributionSummary-sc-11isnv6-2 ccMXsL Details__Summary-xi326o-2 goqLfs"
  >
    <svg
      aria-hidden="true"
      className="Attribution__SummaryClosedIcon-sc-11isnv6-0 gSPJtD Details__ExpandIcon-xi326o-0 cVuUUi StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 192 512"
    >
      <path
        d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        fill="currentColor"
      />
    </svg>
    <svg
      aria-hidden="true"
      className="Attribution__SummaryOpenIcon-sc-11isnv6-1 kRXAve Details__CollapseIcon-xi326o-1 jsdrhW StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 320 512"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
    <span>
      Citation/Attribution
    </span>
  </summary>
  <div
    className="Attribution__Content-sc-11isnv6-3 SNfN"
    dangerouslySetInnerHTML={
      Object {
        "__html": "
<p>
  Want to cite, share, or modify this book? This book is
  <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"http://example.com\\">
    Sweet License 1.0
  </a> and you must attribute Texas Education Agency (TEA). The original material is available at: <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"https://www.texasgateway.org/book/tea-physics\\">
      https://www.texasgateway.org/book/tea-physics
    </a>. Changes were made to the original material, including updates to art, structure, and other content updates.
</p>

<strong>Attribution information</strong>
<ul>
  <li>
    If you are redistributing all or part of this book in a print format,
    then you must include on every physical page the following attribution:
    <blockquote>
      Access for free at https://openstax.org/books/cce64fde-f448-43b8-ae88-27705cceb0da@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units
    </blockquote>
  </li>
  <li>
    If you are redistributing all or part of this book in a digital format,
    then you must include on every digital page view the following attribution:
    <blockquote>
      Access for free at <a href=\\"https://openstax.org/books/cce64fde-f448-43b8-ae88-27705cceb0da@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/cce64fde-f448-43b8-ae88-27705cceb0da@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
    </blockquote>
  </li>
</ul>

<strong>Citation information</strong>
<ul>
  <li>
    Use the information below to generate a citation. We recommend using a
    citation tool such as
    <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"https://www.lib.ncsu.edu/citationbuilder/#/default/default\\">this one</a>.
    <ul>
      <li>
        Authors: Bam Bammerson
      </li>
      <li>
        Publisher/website: OpenStax
      </li>
      <li>
        Book title: Test Book 1
      </li>
      <li>
        Publication date: Jun 21, 2012
      </li>
      <li>
        Location: Houston, Texas
      </li>
      <li>
        Book URL: <a href=\\"https://openstax.org/books/cce64fde-f448-43b8-ae88-27705cceb0da@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/cce64fde-f448-43b8-ae88-27705cceb0da@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
      </li>
      <li>
        Section URL: <a href=\\"https://openstax.org/cool path name\\">https://openstax.org/cool path name</a>
      </li>
    </ul>
  </li>
</ul>

<p>
  © Jun 21, 2012 Texas Education Agency (TEA). <strong>The OpenStax name, OpenStax logo, OpenStax book covers, OpenStax CNX name, and OpenStax CNX logo
  are not subject to the Creative Commons license and may not be reproduced without the prior and express written
  consent of Rice University.</strong>
</p>
",
      }
    }
  />
</details>
`;

exports[`Attribution in browser renders special licensing and attribution for Intellectual Property 1`] = `
<details
  className="Details-xi326o-3 Attribution__AttributionDetails-sc-11isnv6-4 ecFdAU"
  data-analytics-region="attribution"
  data-testid="attribution-details"
>
  <summary
    aria-label="Citation/Attribution"
    className="Attribution__AttributionSummary-sc-11isnv6-2 ccMXsL Details__Summary-xi326o-2 goqLfs"
  >
    <svg
      aria-hidden="true"
      className="Attribution__SummaryClosedIcon-sc-11isnv6-0 gSPJtD Details__ExpandIcon-xi326o-0 cVuUUi StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 192 512"
    >
      <path
        d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        fill="currentColor"
      />
    </svg>
    <svg
      aria-hidden="true"
      className="Attribution__SummaryOpenIcon-sc-11isnv6-1 kRXAve Details__CollapseIcon-xi326o-1 jsdrhW StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 320 512"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
    <span>
      Citation/Attribution
    </span>
  </summary>
  <div
    className="Attribution__Content-sc-11isnv6-3 SNfN"
    dangerouslySetInnerHTML={
      Object {
        "__html": "
<p>
  Want to cite, share, or modify this book? This book is
  <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"http://example.com\\">
    Sweet License 1.0
  </a> and you must attribute The Michelson 20MM Foundation. Changes were made to the original material, including updates to art, structure, and other content updates.
</p>

<strong>Attribution information</strong>
<ul>
  <li>
    If you are redistributing all or part of this book in a print format,
    then you must include on every physical page the following attribution:
    <blockquote>
      Access for free at https://openstax.org/books/1b4ee0ce-ee89-44fa-a5e7-a0db9f0c94b1@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units
    </blockquote>
  </li>
  <li>
    If you are redistributing all or part of this book in a digital format,
    then you must include on every digital page view the following attribution:
    <blockquote>
      Access for free at <a href=\\"https://openstax.org/books/1b4ee0ce-ee89-44fa-a5e7-a0db9f0c94b1@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/1b4ee0ce-ee89-44fa-a5e7-a0db9f0c94b1@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
    </blockquote>
  </li>
</ul>

<strong>Citation information</strong>
<ul>
  <li>
    Use the information below to generate a citation. We recommend using a
    citation tool such as
    <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"https://www.lib.ncsu.edu/citationbuilder/#/default/default\\">this one</a>.
    <ul>
      <li>
        Authors: Bam Bammerson
      </li>
      <li>
        Publisher/website: OpenStax
      </li>
      <li>
        Book title: Test Book 1
      </li>
      <li>
        Publication date: Jun 21, 2012
      </li>
      <li>
        Location: Houston, Texas
      </li>
      <li>
        Book URL: <a href=\\"https://openstax.org/books/1b4ee0ce-ee89-44fa-a5e7-a0db9f0c94b1@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/1b4ee0ce-ee89-44fa-a5e7-a0db9f0c94b1@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
      </li>
      <li>
        Section URL: <a href=\\"https://openstax.org/cool path name\\">https://openstax.org/cool path name</a>
      </li>
    </ul>
  </li>
</ul>

<p>
  © Jun 21, 2012 The Michelson 20MM Foundation. <strong>The OpenStax name, OpenStax logo, OpenStax book covers, OpenStax CNX name, and OpenStax CNX logo
  are not subject to the Creative Commons license and may not be reproduced without the prior and express written
  consent of Rice University.</strong>
</p>
",
      }
    }
  />
</details>
`;

exports[`Attribution in browser renders special licensing and attribution for Statistics 1`] = `
<details
  className="Details-xi326o-3 Attribution__AttributionDetails-sc-11isnv6-4 ecFdAU"
  data-analytics-region="attribution"
  data-testid="attribution-details"
>
  <summary
    aria-label="Citation/Attribution"
    className="Attribution__AttributionSummary-sc-11isnv6-2 ccMXsL Details__Summary-xi326o-2 goqLfs"
  >
    <svg
      aria-hidden="true"
      className="Attribution__SummaryClosedIcon-sc-11isnv6-0 gSPJtD Details__ExpandIcon-xi326o-0 cVuUUi StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 192 512"
    >
      <path
        d="M0 384.662V127.338c0-17.818 21.543-26.741 34.142-14.142l128.662 128.662c7.81 7.81 7.81 20.474 0 28.284L34.142 398.804C21.543 411.404 0 402.48 0 384.662z"
        fill="currentColor"
      />
    </svg>
    <svg
      aria-hidden="true"
      className="Attribution__SummaryOpenIcon-sc-11isnv6-1 kRXAve Details__CollapseIcon-xi326o-1 jsdrhW StyledIconBase-sc-bdy9j4 cCvWXa"
      fill="currentColor"
      focusable="false"
      viewBox="0 0 320 512"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
      />
    </svg>
    <span>
      Citation/Attribution
    </span>
  </summary>
  <div
    className="Attribution__Content-sc-11isnv6-3 SNfN"
    dangerouslySetInnerHTML={
      Object {
        "__html": "
<p>
  Want to cite, share, or modify this book? This book is
  <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"http://example.com\\">
    Sweet License 1.0
  </a> and you must attribute Texas Education Agency (TEA). The original material is available at: <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"https://www.texasgateway.org/book/tea-statistics\\">
      https://www.texasgateway.org/book/tea-statistics
    </a>. Changes were made to the original material, including updates to art, structure, and other content updates.
</p>

<strong>Attribution information</strong>
<ul>
  <li>
    If you are redistributing all or part of this book in a print format,
    then you must include on every physical page the following attribution:
    <blockquote>
      Access for free at https://openstax.org/books/394a1101-fd8f-4875-84fa-55f15b06ba66@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units
    </blockquote>
  </li>
  <li>
    If you are redistributing all or part of this book in a digital format,
    then you must include on every digital page view the following attribution:
    <blockquote>
      Access for free at <a href=\\"https://openstax.org/books/394a1101-fd8f-4875-84fa-55f15b06ba66@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/394a1101-fd8f-4875-84fa-55f15b06ba66@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
    </blockquote>
  </li>
</ul>

<strong>Citation information</strong>
<ul>
  <li>
    Use the information below to generate a citation. We recommend using a
    citation tool such as
    <a target=\\"_blank\\" rel=\\"noopener\\" href=\\"https://www.lib.ncsu.edu/citationbuilder/#/default/default\\">this one</a>.
    <ul>
      <li>
        Authors: Bam Bammerson
      </li>
      <li>
        Publisher/website: OpenStax
      </li>
      <li>
        Book title: Test Book 1
      </li>
      <li>
        Publication date: Jun 21, 2012
      </li>
      <li>
        Location: Houston, Texas
      </li>
      <li>
        Book URL: <a href=\\"https://openstax.org/books/394a1101-fd8f-4875-84fa-55f15b06ba66@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units\\">https://openstax.org/books/394a1101-fd8f-4875-84fa-55f15b06ba66@1.0/pages/1-introduction-to-science-and-the-realm-of-physics-physical-quantities-and-units</a>
      </li>
      <li>
        Section URL: <a href=\\"https://openstax.org/cool path name\\">https://openstax.org/cool path name</a>
      </li>
    </ul>
  </li>
</ul>

<p>
  © Jun 21, 2012 Texas Education Agency (TEA). <strong>The OpenStax name, OpenStax logo, OpenStax book covers, OpenStax CNX name, and OpenStax CNX logo
  are not subject to the Creative Commons license and may not be reproduced without the prior and express written
  consent of Rice University.</strong>
</p>
",
      }
    }
  />
</details>
`;
